tests/sentry/data_export/test_tasks.py

```diff
@@ -51,24 +63,182 @@ def test_issue_by_tag(self):
-            assemble_download(de1.id)
+            assemble_download(de.id)
         error = emailer.call_args[1]["message"]
         assert error == "Requested project does not exist"
-        de2 = ExportedData.objects.create(
+
+    @patch("sentry.data_export.models.ExportedData.email_failure")
+    def test_issue_by_tag_missing_issue(self, emailer):
+        de = ExportedData.objects.create(
             user=self.user,
             organization=self.org,
-            query_type=0,
+            query_type=ExportQueryType.ISSUES_BY_TAG,
             query_info={"project": [self.project.id], "group": -1, "key": "user"},
         )
         with self.tasks():
-            assemble_download(de2.id)
+            assemble_download(de.id)
         error = emailer.call_args[1]["message"]
         assert error == "Requested issue does not exist"
+
+    @patch("sentry.tagstore.get_tag_key")
+    @patch("sentry.utils.snuba.raw_query")
+    @patch("sentry.data_export.models.ExportedData.email_failure")
+    def test_issues_by_tag_outside_retention(self, emailer, mock_query, mock_get_tag_key):
+        """
+        When an issues by tag query goes outside the retention range, it returns 0 results.
+        This gives us an empty CSV with just the headers.
+        """
+        de = ExportedData.objects.create(
+            user=self.user,
+            organization=self.org,
+            query_type=ExportQueryType.ISSUES_BY_TAG,
+            query_info={"project": [self.project.id], "group": self.event.group_id, "key": "foo"},
+        )
+
+        mock_query.side_effect = QueryOutsideRetentionError("test")
+        with self.tasks():
+            assemble_download(de.id)
+        de = ExportedData.objects.get(id=de.id)
+        assert de.date_finished is not None
+        assert de.date_expired is not None
+        assert de.file is not None
+        assert isinstance(de.file, File)
+        assert de.file.headers == {"Content-Type": "text/csv"}
+        # Convert raw csv to list of line-strings
+        header = de.file.getfile().read().strip()
+        assert header == "value,times_seen,last_seen,first_seen"
+
+    @patch("sentry.data_export.models.ExportedData.email_failure")
+    def test_discover(self, emailer):
+        de = ExportedData.objects.create(
+            user=self.user,
+            organization=self.org,
+            query_type=ExportQueryType.DISCOVER,
+            query_info={"project": [self.project.id], "field": ["title"], "query": ""},
+        )
+        with self.tasks():
+            assemble_download(de.id)
+        de = ExportedData.objects.get(id=de.id)
+        assert de.date_finished is not None
+        assert de.date_expired is not None
+        assert de.file is not None
+        assert isinstance(de.file, File)
+        assert de.file.headers == {"Content-Type": "text/csv"}
+        # Convert raw csv to list of line-strings
+        header, raw1, raw2, raw3 = de.file.getfile().read().strip().split("\r\n")
```

Comment:
This line right here has a tendency to fail locally for me if i don't add a sleep before `assemble_download`. Might be due to the event not persisted to clickhouse yet. However, there is a similar test above `test_issue_by_tag` that uses `sentry.tagstore` rather than `sentry.scuba.discover.query` (both calls `raw_query`) seems to work consistently in my experience. Perhaps there is something different there?

Reply:
Turns out, the solution is to set the timestamp of the event to a minute earlier.